layers/ip6.go

```diff
@@ -563,29 +608,85 @@ type IPv6Routing struct {
 // LayerType returns LayerTypeIPv6Routing.
 func (i *IPv6Routing) LayerType() gopacket.LayerType { return LayerTypeIPv6Routing }
 
-func decodeIPv6Routing(data []byte, p gopacket.PacketBuilder) error {
-	base, err := decodeIPv6ExtensionBase(data, p)
+// SerializeTo implementation according to gopacket.SerializableLayer
+func (i *IPv6Routing) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.SerializeOptions) error {
+	bytes, err := b.PrependBytes(8 + len(i.SourceRoutingIPs)*16)
 	if err != nil {
 		return err
 	}
-	i := &IPv6Routing{
-		ipv6ExtensionBase: base,
-		RoutingType:       data[2],
-		SegmentsLeft:      data[3],
-		Reserved:          data[4:8],
+	if opts.FixLengths {
+		i.HeaderLength = uint8(len(i.SourceRoutingIPs) * 2)
 	}
+	bytes[0] = byte(i.NextHeader)
+	bytes[1] = i.HeaderLength
+	bytes[2] = i.RoutingType
+	bytes[3] = i.SegmentsLeft
 	switch i.RoutingType {
-	case 0: // Source routing
-		if (i.ActualLength-8)%16 != 0 {
-			return fmt.Errorf("Invalid IPv6 source routing, length of type 0 packet %d", i.ActualLength)
+	case 0:
+		copy(bytes[4:], i.Reserved)
+	case 4:
+		if opts.FixLengths {
+			i.LastEntry = uint8(len(i.SourceRoutingIPs))
 		}
-		for d := i.Contents[8:]; len(d) >= 16; d = d[16:] {
-			i.SourceRoutingIPs = append(i.SourceRoutingIPs, net.IP(d[:16]))
+		bytes[4] = i.LastEntry
+		bytes[5] = i.Flags
+		binary.BigEndian.PutUint16(bytes[6:], i.Tags)
+	default:
+		return fmt.Errorf("IPv6Routing: RoutingType %d not supported", i.RoutingType)
+	}
+
+	for j, ip := range i.SourceRoutingIPs {
+		copy(bytes[8+j*16:], ip.To16())
+	}
+
+	return nil
+}
+
+// DecodeFromBytes implementation according to gopacket.DecodingLayer
+func (i *IPv6Routing) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	var err error
+	i.ipv6ExtensionBase, err = decodeIPv6ExtensionBase(data, df)
+	if err != nil {
+		return err
+	}
+	if len(data) < 8 {
+		return fmt.Errorf("IPv6Routing: data too short")
+	}
+	i.NextHeader = IPProtocol(data[0])
+	i.HeaderLength = data[1]
+	i.RoutingType = data[2]
+	i.SegmentsLeft = data[3]
+	if len(data)-8 != int(i.HeaderLength)*8 {
+		return fmt.Errorf("IPv6Routing: data length mismatch")
+	}
+	switch i.RoutingType {
+	case 0:
+		i.Reserved = data[4:8]
+	case 4:
+		i.LastEntry = data[4]
+		i.Flags = data[5]
+		i.Tags = binary.BigEndian.Uint16(data[6:8])
+		if len(data)-8 < int(i.LastEntry)*16 {
+			return fmt.Errorf("IPv6Routing: data too short")
 		}
 	default:
-		return fmt.Errorf("Unknown IPv6 routing header type %d", i.RoutingType)
+		return fmt.Errorf("IPv6Routing: RoutingType %d not supported", i.RoutingType)
 	}
+	i.SourceRoutingIPs = make([]net.IP, i.LastEntry)
+	for j := 0; j < int(i.LastEntry); j++ {
```

Comment:
```suggestion
	for j := 0; j <= int(i.LastEntry); j++ {
```
Idem: indexes not sizes.